pyro/infer/util.py

```diff
@@ -39,21 +50,36 @@ def torch_backward(x):
 
 def reduce_to_target(source, target):
     """
-    Sums out any dimensions in source that are of size > 1 in source but of size 1 in target.
-    This preserves source.dim().
+    Sums out any dimensions in source that are of size > 1 in source but of
+    size 1 in target.
     """
-    if source.dim() > target.dim():
-        raise ValueError
+    while source.dim() > target.dim():
+        source = source.sum(0)
     for k in range(1, 1 + source.dim()):
         if source.size(-k) > target.size(-k):
             source = source.sum(-k, keepdim=True)
     return source
 
 
+def reduce_to_shape(source, shape):
```

Comment:
since you've introduced this why not rewrite reduce_to_target in terms of reduce_to_shape?

Reply:
Will do in a follow-up PR